conftest.py

```diff
@@ -0,0 +1,175 @@
+        ' large data files.')
+
+
+def pytest_collection_modifyitems(config, items):
+    r"""
+    Decide which tests to skip based on command-line options.
+    """
+    # Set up the skip marks
+    skip_answer = pytest.mark.skip(reason="--with-answer-testing not set.")
+    skip_big = pytest.mark.skip(reason="--answer-big-data not set.")
+    # Loop over every collected test function
+    for item in items:
+        # If it's an answer test and the appropriate CL option hasn't
+        # been set, skip it
+        if "answer_test" in item.keywords and not config.getoption("--with-answer-testing"):
+            item.add_marker(skip_answer)
+        # If it's an answer test that requires big data and the CL
+        # option hasn't been set, skip it
+        if "big_data" in item.keywords and not config.getoption("--answer-big-data"):
+            item.add_marker(skip_big)
+
+
+@pytest.fixture(scope='function')
+def temp_dir():
+    r"""
+    Creates a temporary directory needed by certain tests.
+    """
+    curdir = os.getcwd()
+    if int(os.environ.get('GENERATE_YTDATA', 0)):
+        tmpdir = os.getcwd()
+    else:
+        tmpdir = tempfile.mkdtemp()
+    os.chdir(tmpdir)
+    yield tmpdir
+    os.chdir(curdir)
+    if tmpdir != curdir:
+        shutil.rmtree(tmpdir)
+
+
+@pytest.fixture(scope='class')
+def answer_file(request):
+    if request.cls.__name__ in pytest.answer_files:
+        answer_file = pytest.answer_files[request.cls.__name__]
+        # Make sure we're not overwriting an existing answer set
+        if os.path.isfile(os.path.join(answer_dir, answer_file)):
+            if request.config.getoption('--answer-store'):
+                raise FileExistsError("Error, attempting to overwrite "
+                    "answer file {}. Either specify a new version or "
+                    "set the `--force-override-answers` option".format(
+                        answer_file
+                    )
+                )
+    else:
+        assert False
+    request.cls.answer_file = answer_file
+
+
+#============================================
+#                param_list
+#============================================
+def param_list(request):
+    r"""
+    Saves the non-ds, non-fixture function arguments for saving to
+    the answer file.
+    """
+    test_params = {}
+    func = request.node.function
+    # co_varnames is all of the variable names local to the function
+    # starting with self, then the passed args, then the vars defined
+    # in the function body. This excludes fixture names
+    args = func.__code__.co_varnames[1:func.__code__.co_argcount]
+    # funcargs includes the names and values of all arguments, including
+    # fixtures, so we use args to weed out the fixtures. Need to have
+    # special treatment of the data files loaded in fixtures for the
+    # frontends
+    for key, val in request.node.funcargs.items():
+        if key in args and not key.startswith('ds_'):
```

Comment:
@ngoldbaum This one has me stumped. I agree that doesn't seem like the best way to do this, but I can't seem to find another way. The purpose of this function is to generate a dictionary of {test_parameter_name : test_parameter_value} pairs that can then be written to the answer file. This allows for easier debugging when a test fails. The crux of the issue is that pytest considers parametrized function arguments as fixtures, so all of the introspecting methods I can find that pytest provides (such as request.node.funcargs) include the non-function-argument parameters (such as hashing and answer_file). I use test parameterization so that each combination of test parameters for a given test would get it's own hash and entry in the answer file. Any thoughts and/or suggestions on fixing this would be greatly appreciated, whenever you get a chance! Thanks!

Reply:
It may be worth asking one of the pytest developers on e.g. their gitter chat. I'm still nervous about this but don't have time to dive in and help you with the code.